Add Freezing Utilities: `freeze`, `selective_optimizer`, and Prefix-Mask Support

pranavagrawaI opened the pull request:
## What’s Included

### 1. New Utilities

- **`freeze(mask)`**  
  Zeroes updates for parameters marked `True` in a (prefix-compatible) boolean mask.

- **`selective_optimizer(optimizer, freeze_mask)`**  
  Applies `optimizer` to parameters where `freeze_mask` is `False`, and freezes the rest via `set_to_zero`.

**Mask semantics:**
- `True`  ⇒ freeze
- `False` ⇒ train
- Missing entries ⇒ default to train

### 2. Prefix-Compatible Mask Expansion

- Implements a helper `_broadcast_mask(params, mask)`:
  - Expands scalar and prefix masks to match full PyTree structure.
  - Supports `dict`, `list`, `tuple`, and array leaves.
  - Differentiates containers vs. scalars safely.
  - Defaults unlabeled branches to `False` (i.e., train).

### 3. Docstrings

- All functions have full docstrings with behavior notes.

## Test Coverage

- Three new test suites `BroadcastMaskTest`, `FreezeTest` and `SelectiveOptimizerTest`.

rdyro commented:
Hey, thanks for the PR! optax should already support [prefix trees](https://docs.jax.dev/en/latest/_autosummary/jax.tree.map.html) for mask (e.g.,` mask = [True, False, True]` and `params = [model1, model2, another_state]` where `model1, model2, another_state` can be nested, so I think no need for custom traversal.

 Do you have an application in mind where the freeze and selective_optimizer utilities could come in handy?

rdyro commented on optax/_src/utils.py:
```diff
@@ -25,12 +25,12 @@
 import jax
 import jax.numpy as jnp
 import jax.scipy.stats.norm as multivariate_normal
+from jax import tree_util
```
jax prefers the full `jax.tree.map` in place of `tree_util.map` - could you please update uses of `tree_util` and remove the import

rdyro commented on optax/_src/utils.py:
```diff
@@ -333,6 +333,178 @@ def _value_and_grad(
+    return False
+
+
+def freeze(mask: base.PyTree) -> base.GradientTransformation:
```
let's rely on jax's prefix tree behavior for jax.tree.map and require explicity prefix/leaf False - for more predictable behavior

rdyro commented on optax/_src/utils.py:
```diff
@@ -333,6 +333,178 @@ def _value_and_grad(
+      An ``optax.GradientTransformation`` that zeroes out gradient updates based
+      on the dynamically broadcasted mask.
+    """
+    from optax.transforms import masked
```
what's the reason for imports not being at the top?

pranavagrawaI replied:
I was facing an issue with cyclic imports when importing at the top, so I moved the imports inside the functions.

Let me know if there is a better way to handle this and I will do that instead.

rdyro commented on optax/_src/utils.py:
```diff
@@ -333,6 +333,178 @@ def _value_and_grad(
+    Returns:
+      An ``optax.GradientTransformation`` that selectively applies the `optimizer`.
+    """
+    from optax.transforms import partition
```
what's the reason for imports not being at the top?

rdyro commented:
I'm worried about maintaining custom traversal logic, so I'd like to avoid it if possible unfortunately. Also, in the spirit of simplicity, we should probably require an explicit prefix `False` for not-/freezing behavior.

Overall great contribution, thanks so much!

pranavagrawaI commented:
I have moved the shortcuts to a `optax/transforms/_freezing.py` and the tests to `optax/transforms/_freezing_test.py`.

Also, do you want me to update the freezing parameters example to also include the new shortcuts or perhaps give the examples in the docstrings?

rdyro commented on optax/transforms/_freezing.py:
```diff
@@ -0,0 +1,93 @@
+        if _is_bool_scalar(mask):
+            return jax.tree.map(lambda _: bool(mask), params)
+
+        chex.assert_trees_all_equal_structs(params, mask)
```
This will likely fail if we're working with a partial prefix mask, right? Can we remove this line?

rdyro commented on optax/transforms/_freezing_test.py:
```diff
@@ -0,0 +1,283 @@
+
+class FreezeTest(parameterized.TestCase):
+
+  @parameterized.named_parameters([
```
Can you add a partial prefix mask as a test case?

The tests have a full prefix mask and a full mask, but we're missing a partial prefix mask like:
```
params = [a, [b, c]]
mask = [False, True]
```

rdyro commented on optax/transforms/_freezing_test.py:
```diff
@@ -0,0 +1,283 @@
+
+class SelectiveOptimizerTest(parameterized.TestCase):
+
+  @parameterized.named_parameters([
```
Like above, we need a partial prefix mask test

rdyro commented on optax/transforms/_freezing_test.py:
```diff
@@ -0,0 +1,283 @@
+  def test_bad_structure_raises(self):
+    bad_mask = {'layer1': [True]}  # missing the bias leaf
+    opt = _freezing.freeze(bad_mask)
+    with self.assertRaises(Exception):
```
Can you make the exception more specific and check for a specific message using the more specific: `self.assertRaisesRegex` please?

[https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex)

rdyro commented on optax/transforms/_freezing_test.py:
```diff
@@ -0,0 +1,283 @@
+  def test_selective_bad_structure(self):
+    bad_mask = {'a': True}  # missing 'b'
+    opt = _freezing.selective_optimizer(alias.sgd(1.0), bad_mask)
+    with self.assertRaises(Exception):
```
Can you make the exception more specific, see above

rdyro commented on optax/transforms/_freezing.py:
```diff
@@ -0,0 +1,93 @@
+    """
+    def mask_fn(params: base.PyTree) -> base.PyTree:
+        if _is_bool_scalar(mask):
+            return jax.tree.map(lambda _: bool(mask), params)
```
This will likely throw a concretization error if mask is a tracer.

Because tree.map supports a prefix map, you can rely on jax.tree.map directly in both scalar and nested case.

Does `return masked(base.set_to_zero(), mask)` work in all your test cases? (Instead of creating a `mask_fn`)

rdyro commented:
> I have moved the shortcuts to a `optax/transforms/_freezing.py` and the tests to `optax/transforms/_freezing_test.py`.
> 
> Also, do you want me to update the freezing parameters example to also include the new shortcuts or perhaps give the examples in the docstrings?

Thank you!

Yes, please! I also left some additional review comments, this is looking great!

pranavagrawaI commented:
### The following changes have been made:

- Simplified the implementations of `freeze `and `selective_optimizer`.
- Removed the internal helper function `_is_bool_scalar`.
- Added a new test case specifically for handling partial prefix masks and combine them to form a single test class for each function.
- Tests now use `self.assertRaisesRegex` 
- Updated the documentation for `freeze` and `selective_optimizer` to explicitly state the requirement for static masks (masks not changing during training).
- Included small usage example in the docstrings for both functions.
- Added freeze and selective_optimizer to optax/transforms/__init__.py, to make them accessible directly under the `optax.transforms` namespace.

I thought of adding specific example uses of these util in [examples/freezing_parameters.ipynb](https://github.com/google-deepmind/optax/blob/main/examples/freezing_parameters.ipynb); I have the notebook ready, but decided not to commit it here, as it maybe better to do that in a separate PR, let me know if you want it here instead.

rdyro commented on optax/transforms/_freezing.py:
```diff
@@ -0,0 +1,86 @@
+          • a PyTree of booleans matching the structure of the parameters, where
+            each leaf indicates whether that specific parameter leaf should be
+            frozen (True) or left unchanged (False).
+
```
Can you add a `.. seealso::` linking to the other function? (an example here: [https://github.com/google-deepmind/optax/blob/5d86211f737a9fe7882b6e22266e9910f3112185/optax/_src/linesearch.py#L245](https://github.com/google-deepmind/optax/blob/5d86211f737a9fe7882b6e22266e9910f3112185/optax/_src/linesearch.py#L245))

pranavagrawaI replied:
a .. seealso:: has been added to both freeze and selectve_transform (previously selective_optimizer)

rdyro commented on optax/transforms/_freezing.py:
```diff
@@ -0,0 +1,86 @@
+) -> "base.GradientTransformation":
+    """
+        Partition updates so that only un-frozen parameters are optimized.
+
```
Can you add a .. seealso:: linking to the other function?

pranavagrawaI replied:
a .. seealso:: has been added to both freeze and selectve_transform (previously selective_optimizer)

rdyro commented on optax/transforms/_freezing.py:
```diff
@@ -0,0 +1,86 @@
+    return masked(base.set_to_zero(), mask)
+
+
+def selective_optimizer(
```
Can we rename to `selective_transform` instead of optimizer, please?

pranavagrawaI replied:
named changed to selective_transform

rdyro reviewed (approved):
Great contribution, thanks!

pranavagrawaI commented:
Happy to Contribute! Just one more thing, it says that one internal check has failed, but I seem to be unable to see the details, let me know if there is a problem and I can fix it.

rdyro commented:
> Happy to Contribute! Just one more thing, it says that one internal check has failed, but I seem to be unable to see the details, let me know if there is a problem and I can fix it.

We have some internal CI that's somewhat difficult to mirror in the open-source github, I have to resolve these internally. It's usually small changes.

rdyro commented:
Can you squash your commits please? We're having some trouble merging the chain internally @pranavagrawaI
